python/ray/_raylet.pyx

```diff
@@ -2726,6 +2730,52 @@ cdef void terminate_asyncio_thread() nogil:
         core_worker = ray._private.worker.global_worker.core_worker
         core_worker.stop_and_join_asyncio_threads_if_exist()
 
+cdef void _cleanup_actor_instance_wrapper() noexcept nogil:
+    """C++ wrapper function that calls the Python actor cleanup callback."""
+    with gil:
+        try:
+            _cleanup_actor_instance()
+        except:
+            # Swallow all exceptions to prevent them from propagating to C++
+            pass
+
+def _cleanup_actor_instance():
+    """Cleanup callback that calls actor's __del__ method."""
+    worker = ray._private.worker.global_worker
+    core_worker = worker.core_worker
+    
+    # Get the actor instance from worker.actors
+    actor_id = core_worker.get_actor_id()
+    if actor_id.is_nil():
+        return
+        
+    actor_instance = worker.actors.get(actor_id)
+    
+    if actor_instance is not None:
+        try:
+            # Call the actor's __del__ method if it exists
+            if hasattr(actor_instance, '__del__'):
+                actor_instance.__del__()
+                logger.debug(f"Actor {actor_id} cleanup completed")
+        except Exception as e:
+            logger.error(f"Error during actor cleanup: {e}")
+        finally:
+            # Remove from actors dict to prevent double cleanup
+            worker.actors.pop(actor_id, None)
```

Comment:
I think this may actually trigger the `__del__` method to be called twice.

In Python, objects are reference counted and `__del__` is called immediately when they go out of scope if they are not part of a reference cycle. Assuming the actor obj is not in a reference cycle, that means that executing this line will cause the actor object to actually be deleted and therefore the `__del__` method to be called.

Reply:
no need to worry about this as per new approach. Also, `actor_cleanup_callback_` is set to `nullptr` after it runs to avoid double execution